# What kind of book should we sell?

We are a new publisher looking to make our first published book a huge success. We focus on our first book because resources are limited and we can't afford to test the market by publishing multiple books and see what would sell the most. Another important detail is we're limited to publishing our books as physical books and e-books; hence, we won't be considering other medium such as audiobook.

Our aim is somewhat modest and we are only looking to be included in the Amazon best sellers list. If we can be included in that list, we will be more likely to be included in other best sellers list such as the New York Times. 

In relation to the Amazon Book Best Sellers list, we want to answer the following questions:
1. Which genre is the most popular?
2. What is the average price of books?
3. How many reviews does a book receive on average? Does the average number of reviews change significantly depending on the genre?
4. How high is the rating a book receives on average? Does the average rating change significantly depending on the genre?
5. Are there any other factors aside from genre that makes a book popular?

We'll start by answering the above plus some incidental questions found along the way; we will the follow up by formulating one or two book profiles that could potentially become best sellers.

## Loading, Cleaning and Exploring the Data

### Load Data Set

In [1]:
import pandas as pd

amazon_books = pd.read_csv("Amazon_popular_books_dataset.csv")
print(amazon_books.shape)

# Display both head and tail of data frame
amazon_books

(2269, 40)


asin         ISBN10  answered_questions availability  \
0     0007350813     0007350813                   0    In Stock.   
1     0007513763  9780007513765                   0    In Stock.   
2     0008183988     0008183988                   0          NaN   
3     0008305838     0008305838                   0    In Stock.   
4     0008375526     0008375526                   0    In Stock.   
...          ...            ...                 ...          ...   
2264  B07P5BPVGM            NaN                   0          NaN   
2265  B07P5JBCFL            NaN                   0          NaN   
2266  B07NF7DFS2            NaN                   0          NaN   
2267  B07P67N918            NaN                   0          NaN   
2268  B07PFY8K2N            NaN                   0          NaN   

                                                  brand currency  \
0                                          Emily Brontë      USD   
1                                          Drew Daywalt      USD   
2                                      Bernard Cornwell      USD   
3                                        David Walliams      USD   
4                                       Caroline Hirons      USD   
...                                                 ...      ...   
2264                                        Jess Lourey      USD   
2265  Heidi Murkoff (Author, Narrator), Meeghan Hola...      USD   
2266  Clea Shearer (Author, Narrator), Joanna Teplin...      USD   
2267  Lisa Jewell (Author), Tamaryn Payne (Narrator)...      USD   
2268  Elton John (Author, Narrator), Taron Egerton (...      USD   

     date_first_available                                           delivery  \
0                     NaN  ["FREE delivery Tuesday, December 28 if you sp...   
1                     NaN  ["FREE delivery Tuesday, December 28 if you sp...   
2                     NaN  ["FREE delivery January 4 - 10 if you spend $2...   
3                     NaN  ["FREE delivery Tuesday, December 28 if you sp...   
4                     NaN  ["FREE delivery Tuesday, December 28","Or fast...   
...                   ...                                                ...   
2264                  NaN                                                 []   
2265                  NaN                                                 []   
2266                  NaN                                                 []   
2267                  NaN                                                 []   
2268                  NaN                                                 []   

      department description  ...  upc                                   url  \
0            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0007350813   
1            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0007513763   
2            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0008183988   
3            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0008305838   
4            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0008375526   
...          ...         ...  ...  ...                                   ...   
2264         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07P5BPVGM   
2265         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07P5JBCFL   
2266         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07NF7DFS2   
2267         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07P67N918   
2268         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07PFY8K2N   

     video  video_count                                         categories  \
0      NaN            0   ["Books","Literature & Fiction","Genre Fiction"]   
1      NaN            0  ["Books","Children's Books","Literature & Fict...   
2      NaN            0   ["Books","Literature & Fiction","Genre Fiction"]   
3      NaN            0  ["Books","Children's Books","Literature & Fict...   
4      NaN            0  ["Books","Crafts, Hobbies & Home","Home I

2269 rows of the data frame matched the data set [documentation](https://github.com/luminati-io/Amazon-popular-books-dataset#amazon-popular-books-dataset):

> This Amazon dataset contains 2269 best-selling books.

On the other hand, the documentation does not mention the number of columns. We should investigate the column names to make sure nothing odd happened when we loaded the data set.

In [2]:
# Show all columns
pd.DataFrame({'columns': amazon_books.columns})

columns
0                   asin
1                 ISBN10
2     answered_questions
3           availability
4                  brand
5               currency
6   date_first_available
7               delivery
8             department
9            description
10              discount
11                domain
12              features
13           final_price
14                format
15             image_url
16          images_count
17         initial_price
18           item_weight
19          manufacturer
20          model_number
21          plus_content
22    product_dimensions
23                rating
24         reviews_count
25          root_bs_rank
26             seller_id
27           seller_name
28             timestamp
29                 title
30                   upc
31                   url
32                 video
33           video_count
34            categories
35     best_sellers_rank
36         buybox_seller
37                 image
38     number_of_sellers
39                colors

### Column Selection

Observing the above results, the column names seems reasonable. Now, we'll need to select the relevant columns because even if keeping irrelevant columns don't have any impact in the results of our analysis, it does clutter the display and also slows down data processing and calculations. 

The documentation mentions that the following are the key columns:

> Key data points included in this free dataset:
> - ASIN
> - ISBN10
> - Categories
> - Reviews count
> - Avg. rating
> - Number of sellers
> - URL
> - Image
> - Final price
> - Title
> - Description
> - Availability

The above columns should be enough for the purpose of this project thus we will restrict our data frame to only those columns for now; we can always add the removed columns later if there's a need for them.

| Full Name         | Column Name       |
|:------------------|:------------------|
| ASIN              | asin              |
| ISBN10            | ISBN10            |
| Categories        | categories        |
| Reviews count     | reviews_count     |
| Average rating    | rating            |
| Number of sellers | number_of_sellers |
| URL               | url               |
| Image             | image             |
| Final price       | final_price       |
| Title             | title             |
| Description       | description       |
| Availability      | availability      |
| Format            | format            |
| Author            | brand             |

We've also added the columns for Format and Author (which is called `brand` - odd, but makes sense because sometimes the book is authored collectively as an institution or publisher; and yes, authors are themselves brands in a way).

In [3]:
# Define the columns to keep
keep_cols = ['asin', 
             'ISBN10', 
             'categories', 
             'reviews_count', 
             'rating', 
             'number_of_sellers', 
             'url', 
             'image', 
             'final_price',
             'title',
             'description',
             'availability',
             'format',
             'brand'
            ]

# Review
amazon_books_updated = amazon_books[keep_cols]
amazon_books_updated.head()

asin         ISBN10  \
0  0007350813     0007350813   
1  0007513763  9780007513765   
2  0008183988     0008183988   
3  0008305838     0008305838   
4  0008375526     0008375526   

                                          categories  reviews_count  \
0   ["Books","Literature & Fiction","Genre Fiction"]          13451   
1  ["Books","Children's Books","Literature & Fict...          16628   
2   ["Books","Literature & Fiction","Genre Fiction"]          11275   
3  ["Books","Children's Books","Literature & Fict...          15520   
4  ["Books","Crafts, Hobbies & Home","Home Improv...          10884   

               rating  number_of_sellers  \
0  4.6 out of 5 stars                NaN   
1  4.8 out of 5 stars                NaN   
2  4.8 out of 5 stars                NaN   
3  4.8 out of 5 stars                NaN   
4  4.8 out of 5 stars                NaN   

                                    url image  final_price  \
0  https://www.amazon.com/dp/0007350813   NaN         3.99   
1  https://www.amazon.com/dp/0007513763   NaN        12.08   
2  https://www.amazon.com/dp/0008183988   NaN          NaN   
3  https://www.amazon.com/dp/0008305838   NaN        20.43   
4  https://www.amazon.com/dp/0008375526   NaN        28.89   

                                               title description availability  \
0               Wuthering Heights (Collins Classics)         NaN    In Stock.   
1                          THE DAYS THE CRAYONS QUIT         NaN    In Stock.   
2        War Lord: Book 13 (The Last Kingdom Series)         NaN          NaN   
3  Code Name Bananas: The hilarious and epic new ...         NaN    In Stock.   
4  Skincare: The award-winning ultimate no-nonsen...         NaN    In Stock.   

                                              format             brand  
0  [{"name":"Kindle","price":"$0.99","url":"/Wuth...      Emily Brontë  
1  [{"name":"Kindle","price":"$10.99","url":"/Day...      Drew Daywalt  
2  [{"name":"Kindle","price":"$11.99","url":"/War...  Bernard Cornwell  
3  [{"name":"Paperback","price":"$24.78","url":"/...    David Walliams  
4  [{"name":"Kindle","price":"$16.99","url":"/Ski...   Caroline Hirons

### Another Round of Column Selection

The previous column selection is mostly done to retrieve plausible columns for the analysis. Upon closer inspection, there are some columns I would like to remove:

| Full Name    | Column Name  | Reason for Removal                                                                                                                                                                                                                                                                                                                                                                                         |
|:-------------|:-------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ASIN         | asin         | The [Amazon Standard Identification Number](https://en.wikipedia.org/wiki/Amazon_Standard_Identification_Number) (ASIN) is not required. It's sufficient to identify a book by its title.                                                                                                                                                                                                                  |
| ISBN         | ISBN10       | Similar to the reason for ASIN removal, the [International Standard Book Number](https://en.wikipedia.org/wiki/ISBN) (ISBN) is not required.                                                                                                                                                                                                                                                               |
| URL          | url          | The url or link to the Amazon page for the book is not necessary. Our analysis doesn't cover anything that looks into the exact url or the page the url links to.                                                                                                                                                                                                                                          |
| Image        | image        | The image of the book is not needed. It could be useful, but it requires different (and potentially more complicated) data analysis methods which are beyond the scope of this project. A worthy candidate for investigation in our next project.                                                                                                                                                          |
| Description  | description  | Similar to book image, this could be useful but can be too complicated to analyse for the purpose of this project.                                                                                                                                                                                                                                                                                         |
| Availability | availability | We are not concerned with a book's stock availability on Amazon; if the book is included in the best seller list, it's good enough whether it's available or not. Other than that, stock availability data is dynamic and the data set is just a snapshot of a book's availability when the data was collected; we'll need a time series data set of availability if we want to use the data for analysis. |

In [4]:
# Define the columns to remove
remove_cols =   [
                 'asin', 
                 'ISBN10', 
                 'url', 
                 'image', 
                 'description',
                 'availability',
                 ]

# Remove the columns
amazon_books_updated = amazon_books_updated.drop(columns=remove_cols)

# Review
amazon_books_updated.head()

categories  reviews_count  \
0   ["Books","Literature & Fiction","Genre Fiction"]          13451   
1  ["Books","Children's Books","Literature & Fict...          16628   
2   ["Books","Literature & Fiction","Genre Fiction"]          11275   
3  ["Books","Children's Books","Literature & Fict...          15520   
4  ["Books","Crafts, Hobbies & Home","Home Improv...          10884   

               rating  number_of_sellers  final_price  \
0  4.6 out of 5 stars                NaN         3.99   
1  4.8 out of 5 stars                NaN        12.08   
2  4.8 out of 5 stars                NaN          NaN   
3  4.8 out of 5 stars                NaN        20.43   
4  4.8 out of 5 stars                NaN        28.89   

                                               title  \
0               Wuthering Heights (Collins Classics)   
1                          THE DAYS THE CRAYONS QUIT   
2        War Lord: Book 13 (The Last Kingdom Series)   
3  Code Name Bananas: The hilarious and epic new ...   
4  Skincare: The award-winning ultimate no-nonsen...   

                                              format             brand  
0  [{"name":"Kindle","price":"$0.99","url":"/Wuth...      Emily Brontë  
1  [{"name":"Kindle","price":"$10.99","url":"/Day...      Drew Daywalt  
2  [{"name":"Kindle","price":"$11.99","url":"/War...  Bernard Cornwell  
3  [{"name":"Paperback","price":"$24.78","url":"/...    David Walliams  
4  [{"name":"Kindle","price":"$16.99","url":"/Ski...   Caroline Hirons

### Handling missing values

Let's check for missing values.

In [5]:
# Show the number of missing values per column
amazon_books_updated.isnull().sum()

categories              0
reviews_count           0
rating                  0
number_of_sellers    2265
final_price           877
title                   0
format                 96
brand                   1
dtype: int64

It seems that the number of sellers, final price, format, and brand columns have missing values. The first two, especially  number of sellers, have startling number of missing values. Next, I'll go through each column one by one to evaluate how to handle them.

**Missing values in the number of sellers column**

There's a significant number of missing values in this column. Interesting enough, there are 3 rows with a value.

In [6]:
# Show frequency of values in number_of_sellers
amazon_books_updated['number_of_sellers'].value_counts()

# Show rows has values in number_of_sellers
amazon_books_updated[amazon_books_updated['number_of_sellers'].notnull()]

categories  reviews_count  \
664   ["Books","Humor & Entertainment","Puzzles & Ga...          37782   
1194  ["Books","Self-Help","Relationships","Love & R...          15377   
2029  ["Toys & Games","Learning & Education","Electr...          18477   
2177             ["Books","Calendars","Animals","Dogs"]          11249   

                  rating  number_of_sellers  final_price  \
664   4.9 out of 5 stars               94.0        27.93   
1194  4.7 out of 5 stars               19.0         8.01   
2029  4.8 out of 5 stars                4.0        17.99   
2177  4.8 out of 5 stars                2.0        16.99   

                                                  title format  \
664              Player's Handbook (Dungeons & Dragons)    NaN   
1194  Knock Knock What I Love about You Fill in the ...    NaN   
2029                    VTech Musical Rhymes Book, Pink    NaN   
2177   Pooping Pooches White Elephant Gag Gift Calendar    NaN   

                     brand  
664   Wizards of the Coast  
1194           Knock Knock  
2029                 VTech  
2177       Pooping Pooches

It doesn't seem that we can glean anything important from the available values. Doing the work ourself by filling in the number of sellers is not an option because we do not know exactly when the data was collected; even if we do know, it's unclear whether it's possible to find past data for the known date. Our only option is to drop the `number_of_sellers` though not reluctantly because it is not too important for our analysis.

In [7]:
amazon_books_updated.drop(columns=['number_of_sellers'], inplace=True)
# Verify column has been dropped
pd.DataFrame({"columns": amazon_books_updated.columns})

columns
0     categories
1  reviews_count
2         rating
3    final_price
4          title
5         format
6          brand

**Missing values in the final price and format columns**

We put both the final price and format columns together in this section because they are actually related. Let's see what we mean by that:

In [8]:
# Print the frequency of values in both the final price and format columns
print(amazon_books_updated['final_price'].value_counts(dropna=False))

# Be ready for an eyesore
print(amazon_books_updated['format'].value_counts(dropna=False).head())

NaN      877
9.99      25
6.99      24
8.99      23
15.99     21
        ... 
17.26      1
25.01      1
21.06      1
58.88      1
9.48       1
Name: final_price, Length: 733, dtype: int64
NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                96
[{"name":"Kindle","price":"$0.00","url":"/Picture-Dorian-Gray-AmazonClassics-ebook/dp/B071HDXV91"},{"n

The values in the final price column is as expected, but the values in the format column needs a further inspection. Before that, we need to some data conversion because even though the above looks like a list, it is actually a string. Converting all the values into a list is our next step.

In [12]:
import ast
import numpy as np

'''
**Convert each value that is not null into a list**

`ast.literal_eval` evaluates and converts the literal string into the appropriate Python object. 

The [documentation](https://docs.python.org/3/library/ast.html#ast.literal_eval) mentioned that it is safe-ish; it prevents Python code from being called but depending on the input, it's possible to trigger memory or C-stack exhaustion.

We trust the collector of the data set enough to use `ast.literal_eval` in spite of the potential attack.
'''
amazon_books_updated['format'] = amazon_books_updated['format'].apply(lambda x: ast.literal_eval(x) if not pd.isna(x) else np.nan)

# Verify if the number of nulls is still the same
amazon_books_updated['format'].isnull().sum()

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

Next: let's check what formats and price are used.

In [10]:
'''
The value in the format column is a list containing dictionaries. 

In each list is a dictionary that follows the following structure:
{
    "name": "Format",
    "price": "$0.00",
    "url": "/url/to/the/book/page/on/Amazon"
}

We would like to extract the name and price for each book. 
But before that, let's see the possible values for the name and price.
''' 


# for formats in test:
#     print(type(formats))
# for format in amazon_books_updated[amazon_books_updated['format'].isna()]:
#     print(formats)
# amazon_books_updated[amazon_books_updated['format'].isna()]
# for formats in amazon_books_updated['format']:
#     print(formats)
# for formats in amazon_books_updated['format']:
#     print(ast.literal_eval(formats))
#     print(type(formats))
#     for single_format in formats:
#         print(single_format)
#         format_list.add(single_format['name'])
#         price_list.add(single_format['price'])

# price_list = {book_format['price'] for book_format in amazon_books_updated['format']}
# format_list = {book_format['name'] for book_format in amazon_books_updated['format']}

'\nThe value in the format column is a list containing dictionaries. \n\nIn each list is a dictionary that follows the following structure:\n{\n    "name": "Format",\n    "price": "$0.00",\n    "url": "/url/to/the/book/page/on/Amazon"\n}\n\nWe would like to extract the name and price for each book. \nBut before that, let\'s see the possible values for the name and price.\n'